x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,1244 @@
+    {
+        // If we need to drive the connection, we need to acquire the reader lock.
+        if let Some(packet_reader) = self.packet_reader.try_lock() {
+            // Run the future and drive at the same time.
+            future::or(future, async move {
+                self.drive(packet_reader).await.map(|x| match x {})
+            })
+            .await
+        } else {
+            // Someone else is driving, so just run the future.
+            future.await
+        }
+    }
+
+    async fn drive(
+        &self,
+        mut packet_reader: MutexGuard<'_, PacketReader>,
+    ) -> Result<Infallible, ConnectionError> {
+        let mut fds = vec![];
+        let mut packets = vec![];
+
+        loop {
+            for _ in 0..50 {
+                // Try to read packets from the stream.
+                packet_reader.try_read_packets(&self.stream, &mut packets, &mut fds)?;
+                let packet_count = packets.len();
+
+                // Now, actually enqueue the packets.
+                {
+                    let mut inner = self.inner.lock().unwrap();
+                    inner.conn.enqueue_fds(mem::take(&mut fds));
+                    packets
+                        .drain(..)
+                        .for_each(|packet| inner.conn.enqueue_packet(packet));
+                }
+
+                if packet_count > 0 {
+                    // Notify any listeners that there is new data.
+                    self.new_input.notify(1);
+                } else {
+                    // Wait for more data.
+                    self.stream.readable().await?;
+                }
+            }
+
+            // In the case of a large influx of packets, don't starve other tasks.
+            future::yield_now().await;
+        }
+    }
+
+    async fn prefetch_extension_impl(
+        &self,
+        ext: &'static str,
+    ) -> Result<RwLockWriteGuard<'_, Extensions>, ConnectionError>
+    where
+        S: Send + Sync,
+    {
+        // See if there is an entry in the cache.
+        let mut cache = self.extensions.write().await;
+
+        if !cache.0.contains_key(&ext) {
+            // Check again if there is a cache entry.
+            if let Entry::Vacant(entry) = cache.0.entry(ext) {
+                // Send a QueryExtension request.
+                let cookie = crate::protocol::xproto::query_extension(self, ext.as_bytes()).await?;
+
+                // Add the extension to the cache.
+                entry.insert(ExtensionState::Loading(cookie.sequence_number()));
+
+                std::mem::forget(cookie);
+            }
+
+            return Ok(cache);
+        }
+
+        Ok(cache)
+    }
+
+    /// Prefetch the maximum request length.
+    async fn prefetch_len_impl(
+        &self,
+    ) -> Result<Option<MutexGuard<'_, MaxRequestBytes>>, ConnectionError>
+    where
+        S: Send + Sync,
+    {
+        let mut mrl = match self.max_request_bytes.try_lock() {
```

Comment:
The mutex being locked is taken as a signal that another part of the program is using the current request to find out the maximum bytes.

Reply:
Yeah, but couldn't it be another task that holds this mutex? I don't think using the mutex like this actually works.

So... how does x11rb do it? `RustConnection::maximum_request_bytes()` unconditionally locks the mutex. [This would not work if the current thread already holds the mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.lock).

Also, the code in `compute_length_field` only calls `maximum_request_bytes()` when the request length does not fit into an u16. That feels a bit like a bug (couldn't an X11 server have a maximum request length smaller than 2^18?), but that should mean that all requests from the BigRequests extension do not actually hit this code here...